### PR TITLE
[rllib] Provide internal access to episode state in compute_actions() and allow returning extra batches

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

The goal of this PR is to allow custom policies to perform model-based rollouts. In the multi-agent setting, this requires access to not only policies of other agents, but also their current observations.
Also, you might want to return the model-based trajectories as part of the rollout for efficiency.

- [x] compute_actions() now takes a new keyword arg `episodes`
- [x] pull out internal episode class into a top-level file
- [x] add function to return extra trajectories from an episode that will be appended to the sample batch
- [x] documentation

### CODE REVIEW

File: python/ray/rllib/evaluation/episode.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+from __future__ import absolute_import
```
Comment (ericl) [inline python/ray/rllib/evaluation/episode.py]: Moved from sampler.py to its own file

File: python/ray/rllib/evaluation/tf_policy_graph.py [hunk 2]
```diff
@@ -35,6 +35,10 @@ class TFPolicyGraph(PolicyGraph):
         SampleBatch({"action": ..., "advantages": ..., ...})
     """
 
+    # Set this to False if you override compute_actions() only, and do not
+    # implement build_compute_actions().
+    _supports_build_compute_actions = True
```
Comment (ericl) [inline python/ray/rllib/evaluation/tf_policy_graph.py]: Couldn't think of a better way to handle this...

**Comment by ericl** [comment]
cc @eringrant @richardliaw

File: python/ray/rllib/evaluation/episode.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+class MultiAgentEpisode(object):
+    """Tracks the current state of a (possibly multi-agent) episode.
+
+    The APIs in this class should be considered experimental, but we should
+    avoid changing things for the sake of changing them since users may
+    depend on them for advanced algorithms.
+
+    Attributes:
+        new_batch_builder (func): Create a new empty MultiAgentBatchBuilder.
```
Comment (eringrant) [inline python/ray/rllib/evaluation/episode.py]: `MultiAgentBatch`

Comment (eringrant) [inline python/ray/rllib/evaluation/episode.py]: `MultiAgentSampleBatchBuilder`

Comment (ericl) [reply python/ray/rllib/evaluation/episode.py]: Changed

**Comment by ericl** [comment]
also cc @cathywu @eugenevinitsky if you're interested

File: python/ray/rllib/evaluation/episode.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            policy = self._policies[self.policy_for(agent_id)]
+            self._agent_to_rnn_state[agent_id] = policy.get_initial_state()
+        return self._agent_to_rnn_state[agent_id]
+
+    def last_pi_info_for(self, agent_id):
+        """Returns the last info object for the specified agent."""
+
+        return self._agent_to_last_pi_info[agent_id]
+
+    def add_agent_rewards(self, reward_dict):
```
Comment (richardliaw) [inline python/ray/rllib/evaluation/episode.py]: should docstring all public methods now that this is more user-exposed

**Review by richardliaw — APPROVED** [review]

File: python/ray/rllib/evaluation/sample_batch.py [hunk 5]
```diff
@@ -117,6 +117,11 @@ def postprocess_batch_so_far(self):
             other_batches = pre_batches.copy()
             del other_batches[agent_id]
             policy = self.policy_map[self.agent_to_policy[agent_id]]
+            if True in pre_batch["dones"][:-1] or len(
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/evaluation/sample_batch.py]: Done

File: python/ray/rllib/evaluation/episode.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/evaluation/episode.py]: Made these private

**Pull request merged**